### DISCUSSION TITLE
vips_thumbnail_image is significantly slower than vips_thumbnail_buffer for webp

### DISCUSSION THREAD

**Discussion by rtio** [libvips/libvips#3375] [Q&A] [answered]
I'm currently working with vips using CGO, and I've come across an issue that's puzzling me. When I call certain vips functions, everything runs smoothly:

- `vips_thumbnail_buffer` - I read my buffer using
- `vips_webpsave_buffer` - After making some changes, I save it back into a buffer

However, I recently changed the structure of my code and had to switch from using `vips_webpsave_buffer` to creating a `VipsImage` from `vips_webpload_buffer`. The new process looks like this:

- `vips_webpload_buffer` - I read my buffer
- `vips_thumbnail_image` - I resize the VipsImage
- `vips_webpsave_buffer` - I convert the VipsImage back to a buffer

Unfortunately, this new process is significantly slower than the previous one. I'm wondering if there's anything I can do to maintain my current approach without sacrificing performance.

**Comment by jcupitt** [comment, marked as answer]
Hi @rtio,

Yes, `thumbnail_image` is for emergencies only, it's a lot slower.

`thumbnail`, `thumbnail_buffer` and `thumbnail_source` combine open and resize in one operation, so they can exploit a lot of tricks in the load libraries. You can expect maybe a 10x slowdown for some formats, and a loss in quality for others.

**Reply by rtio** [reply]
Thanks for the answer. I saw somewhere the speed increase would only infer over jpg, but I was wrong, due to that I give it a try as I'm only handling WebP, may I ask if there is an alternative or even a way to mimic the results of thumbnail_buffer when handling with VipsImage?

**Reply by jcupitt** [reply]
Lots of libraries have load format tricks. You **must** design your image handling to have open and resize in the same place to be able to take advantage of these things, and I don't think there's any real way around that. 

For formats like SVG and PDF you'll see a dramatic improvement in quality as well, since they can directly render the object to the required size.

`thumbnail_image` is only for emergencies, and can't do any of this clever stuff.

**Reply by rtio** [reply]
Thank you for answering this. I'll take another approach.
